islandora_compound_object.module

```diff
@@ -268,13 +268,58 @@ function islandora_compound_object_access(AbstractObject $object) {
 }
 
 /**
- * Replaces parent object thumbnail with child's thumbnail.
+ * Implements hook_compound_cmodel_islandora_derivative().
  */
-function islandora_compound_object_update_parent_thumbnail($parent) {
+function islandora_compound_object_compoundCModel_islandora_derivative(AbstractObject $object = NULL, $ds_modified_params = array()) {
+  $module_path = drupal_get_path('module', 'islandora_solution_pack_compound');
+  $derivatives = array();
+  if (variable_get('islandora_compound_object_tn_deriv_hooks', FALSE)) {
+    $derivatives[] = array(
+      'source_dsid' => NULL,
+      'destination_dsid' => 'TN',
+      'function' => array('islandora_compound_object_create_thumbnail'),
+      'file' => "$module_path/includes/derivatives.inc",
+    );
+  }
+
+  return $derivatives;
+
+}
+
+/**
+ * Implements hook_islandora_derivative().
+ */
+function islandora_compound_object_islandora_derivative(AbstractObject $object) {
+  $module_path = drupal_get_path('module', 'islandora_solution_pack_compound');
+  $derivatives = array();
+  // Short circuit if feature is disabled.
+  if (variable_get('islandora_compound_object_tn_deriv_hooks', FALSE)) {
+    $derivatives[] = array(
+      'source_dsid' => 'TN',
+      'destination_dsid' => NULL,
+      'function' => array('islandora_compound_object_create_thumbnail_on_parent'),
+      'file' => "$module_path/includes/derivatives.inc",
+    );
+  }
+  return $derivatives;
+}
+
+/**
+ * Update a parent object's TN by deriving it's first child's TN.
+ *
+ * @param AbstractObject $parent
+ *   The compound object.
+ * @param AbstractObject $child
+ *   The child object (optional).
+ */
+function islandora_compound_object_update_parent_thumbnail(AbstractObject $parent, AbstractObject $child = NULL) {
   module_load_include('inc', 'islandora', 'includes/utilities');
+
   $parts = islandora_compound_object_get_parts($parent->id);
```

Comment:
`islandora_compound_object_get_parts()` is rather heavy; if there's a chance it's not needed (i.e., `$child` is already set), could we skip it?